In [2]:
import tensorflow as tf
from tensorflow.keras.datasets import mnist
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Flatten
from tensorflow.keras.utils import to_categorical

# 1. Load dataset MNIST
(X_train, y_train), (X_test, y_test) = mnist.load_data()

# Normalisasi data (0-255 → 0-1)
X_train = X_train / 255.0
X_test = X_test / 255.0

# One-hot encoding label
y_train = to_categorical(y_train, 10)
y_test = to_categorical(y_test, 10)

# 2. Bangun model JST
model = Sequential([
    Flatten(input_shape=(28, 28)),        # Ubah 28x28 jadi vektor 784
    Dense(128, activation='relu'),        # Hidden layer 1
    Dense(64, activation='relu'),         # Hidden layer 2
    Dense(10, activation='softmax')       # Output layer (10 kelas)
])

# 3. Kompilasi model
model.compile(optimizer='adam',
              loss='categorical_crossentropy',
              metrics=['accuracy'])

# 4. Latih model
model.fit(X_train, y_train, epochs=10, batch_size=32, verbose=1)

# 5. Evaluasi model
loss, acc = model.evaluate(X_test, y_test)
print(f"Akurasi pada data uji: {acc:.4f}")


c:\Users\USER\anaconda3\Lib\site-packages\keras\src\layers\reshaping\flatten.py:37: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Epoch 1/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 6s 2ms/step - accuracy: 0.9301 - loss: 0.2380
Epoch 2/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 4s 2ms/step - accuracy: 0.9692 - loss: 0.1010
Epoch 3/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 4s 2ms/step - accuracy: 0.9779 - loss: 0.0707
Epoch 4/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 4s 2ms/step - accuracy: 0.9830 - loss: 0.0523
Epoch 5/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 4s 2ms/step - accuracy: 0.9857 - loss: 0.0439
Epoch 6/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 4s 2ms/step - accuracy: 0.9883 - loss: 0.0360
Epoch 7/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 4s 2ms/step - accuracy: 0.9903 - loss: 0.0292
Epoch 8/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 4s 2ms/step - accuracy: 0.9917 - loss: 0.0253
Epoch 9/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 4s 2ms/step - accuracy: 0.9934 - loss: 0.0209
Epoch 10/10
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 4s 2ms/step - accuracy: 0.9932 - loss: 0.0193
313/313 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.9779 - loss: 0.0925
Akurasi pada data uji: 0.9779


## Coba dengan beberapa parameter lain:

Ubah jumlah neuron di hidden layer (misal: 256 dan 128).

Tambahkan satu hidden layer lagi.

Bandingkan akurasi dan waktu pelatihan.

Eksperimen dengan fungsi aktivasi Sigmoid vs ReLU.

In [1]:
# ================================
# MNIST MLP EXPERIMENT (FULL CODE)
# ================================

import tensorflow as tf
from tensorflow.keras.datasets import mnist
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Flatten
from tensorflow.keras.utils import to_categorical

# ================================
# 1. Load & Preprocessing Dataset
# ================================
(X_train, y_train), (X_test, y_test) = mnist.load_data()

# Normalisasi pixel 0-255 menjadi 0-1
X_train = X_train / 255.0
X_test = X_test / 255.0

# One-hot encoding pada label
y_train = to_categorical(y_train, 10)
y_test = to_categorical(y_test, 10)

# ============================================
# 2. Fungsi untuk membangun dan evaluasi model
# ============================================
def build_and_train(model):
    model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])
    model.fit(X_train, y_train, epochs=10, batch_size=32, verbose=0)
    loss, acc = model.evaluate(X_test, y_test, verbose=0)
    return acc, loss

# ================================
# 3. EXPERIMENT A: Ubah jumlah neuron
# ================================
model_neuron = Sequential([
    Flatten(input_shape=(28, 28)),
    Dense(256, activation='relu'),
    Dense(128, activation='relu'),
    Dense(10, activation='softmax')
])

acc_neuron, loss_neuron = build_and_train(model_neuron)

# ================================
# 4. EXPERIMENT B: Tambah 1 Hidden Layer
# ================================
model_3layer = Sequential([
    Flatten(input_shape=(28, 28)),
    Dense(256, activation='relu'),
    Dense(128, activation='relu'),
    Dense(64, activation='relu'),  # tambahan layer
    Dense(10, activation='softmax')
])

acc_3layer, loss_3layer = build_and_train(model_3layer)

# ================================
# 5. EXPERIMENT C: Sigmoid vs ReLU
# ================================

# Model dengan SIGMOID
model_sigmoid = Sequential([
    Flatten(input_shape=(28, 28)),
    Dense(256, activation='sigmoid'),
    Dense(128, activation='sigmoid'),
    Dense(64, activation='sigmoid'),
    Dense(10, activation='softmax')
])
acc_sigmoid, loss_sigmoid = build_and_train(model_sigmoid)

# Model dengan ReLU
model_relu = Sequential([
    Flatten(input_shape=(28, 28)),
    Dense(256, activation='relu'),
    Dense(128, activation='relu'),
    Dense(64, activation='relu'),
    Dense(10, activation='softmax')
])
acc_relu, loss_relu = build_and_train(model_relu)

# ================================
# 6. Print Perbandingan Hasil
# ================================
print("\n================ HASIL EKSPEIMEN MNIST =================")
print(f"(A) Ubah Neuron (256,128)     -> Acc: {acc_neuron:.4f} | Loss: {loss_neuron:.4f}")
print(f"(B) Tambah 1 Hidden Layer      -> Acc: {acc_3layer:.4f} | Loss: {loss_3layer:.4f}")
print(f"(C1) SIGMOID (3 Hidden)        -> Acc: {acc_sigmoid:.4f} | Loss: {loss_sigmoid:.4f}")
print(f"(C2) RELU (3 Hidden)           -> Acc: {acc_relu:.4f} | Loss: {loss_relu:.4f}")


c:\Users\USER\anaconda3\Lib\site-packages\keras\src\layers\reshaping\flatten.py:37: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)



================ HASIL EKSPEIMEN MNIST =================
(A) Ubah Neuron (256,128)     -> Acc: 0.9813 | Loss: 0.0839
(B) Tambah 1 Hidden Layer      -> Acc: 0.9785 | Loss: 0.0932
(C1) SIGMOID (3 Hidden)        -> Acc: 0.9791 | Loss: 0.0819
(C2) RELU (3 Hidden)           -> Acc: 0.9831 | Loss: 0.0765


### Ringkasan Singkat Perbandingan MNIST

Model awal dengan dua hidden layer menghasilkan akurasi 0.9779. Ketika jumlah neuron diperbesar menjadi 256 dan 128, akurasi meningkat menjadi 0.9813, namun waktu pelatihannya lebih lama karena komputasi bertambah. Penambahan satu hidden layer juga sedikit meningkatkan akurasi menjadi 0.9785, tetapi proses training menjadi lebih berat.

Perbandingan fungsi aktivasi menunjukkan bahwa Sigmoid memberikan akurasi lebih rendah dan pelatihan lebih lambat dibanding ReLU. ReLU menghasilkan akurasi tertinggi, yaitu 0.9831 pada tiga hidden layer, serta melatih lebih cepat karena tidak mengalami vanishing gradient.

Secara keseluruhan, model dengan tiga hidden layer dan aktivasi ReLU memberikan performa terbaik, baik dari akurasi maupun efisiensi pelatihan.
